src/auth.js

```diff
@@ -44,9 +64,98 @@ function validRedirect(app, path) {
     return false;
 }
 
-function authenticateGoogleIdToken(token) {
+async function authenticateWithGoogleAuth(res, clientRedirect, saveSession) {
+    // relative URLs and those matching certain domains are accepted
+    if (!clientRedirect.startsWith('/') && 
+        (!clientRedirect || !OAUTH_CLIENT_REDIRECT_URIS_REGEX.find(regex => regex.test(clientRedirect)))
+    ) {
+        res.status(400).send('Value of query parameter to is invalid');
+        return;
+    }
+
+    const oAuth2Client = newOauthClient();
+
+    const state = crypto.randomBytes(20).toString('hex');
+    authRequestState[state] = { clientRedirect: clientRedirect, saveSession, created: Date.now() };
+
+    const authorizeUrl = oAuth2Client.generateAuthUrl({
+        access_type: 'offline',
+        scope: 'https://www.googleapis.com/auth/userinfo.profile https://www.googleapis.com/auth/userinfo.email',
+        state,
+        prompt: 'consent',
+    });
+    res.redirect(authorizeUrl);
+}
+
+async function handleGoogleAuthCallback(req, res) {
+    const reqCode = req.query.code;
+    if (!reqCode) {
+        res.status(400).send('Missing required code parameter in Google Auth callback');
+        return;
+    }
+
+    const state = loadAuthRequestStateOrSetResponse(req, res);
+    clearLongOverdueStates();
+    if (!state) {
+        return;
+    }
+
+    const oAuth2Client = newOauthClient();
+    const tokenRes = await oAuth2Client.getToken(reqCode);
+
+    const data = await authenticateGoogleIdToken(tokenRes.tokens.id_token, oAuth2Client);
+    
+    const apiToken = common.auth.signAPIAccessToken({
+        role: process.env.API_ROLE || 'employee',
+        // TODO: Should fetch employee ID instead.
+        email: data.email
+    });
+
+    if (state.saveSession) {
+        req.session.apiToken = apiToken;
+        req.session.email = data.email;
+        // TODO: Supplying google id_token too for now, until all apps are changed over.
+        req.session.id_token = req.body.id_token;
+
+        res.redirect(state.clientRedirect);
+    } else {
+        res.redirect(`${state.clientRedirect}?access_token=${apiToken}&refresh_token=${tokenRes.tokens.refresh_token}`);
+    }
+}
+
+function loadAuthRequestStateOrSetResponse(req, res) {
+    const reqState = req.query.state;
+    if (!reqState) {
+        res.status(400).send('Missing required state parameter in Google Auth callback');
+        return null;
+    }
+    const cachedState = authRequestState[reqState];
+    if (!cachedState) {
+        res.status(400).send('Unknown value in state parameter');
+        return null;
+    }
+    delete authRequestState[reqState];
+    if (Date.now() - cachedState.created > OATH_STATE_TTL) {
+        res.status(400).send(`State has expired, please complete authentication within ${OATH_STATE_TTL / 1000 / 60} minutes`);
+        return null;
+    }
+    return cachedState;
+}
+
+/**
+ * Removes abandoned auth request states
+ */
```

Comment:
Bør kanskje bruke samme kommentar-markører som ellers i koden?

```suggestion
// Removes abandoned auth request states
```